File: scripts/normalise-reports.js
#!/usr/bin/env node
/**
 * DYOR HQ v2 — Report Normalisation Script
 * Ensures every report follows the canonical template structure exactly.
 * 
 * Runs in place — modifies reports/*.html directly.
 * No commit until validated.
 *
 * Canonical sections (in order):
 *  1. Executive Summary
 *  2. Business Model
 *  3. Financial Snapshot
 *  4. Recent Catalysts
 *  5. Thesis Evaluation
 *  6. Key Risks
 *  7. Who Should Own It / Avoid It
 *  8. Recommendation
 *  9. Entry / Exit Framework
 * 10. Sources
 *
 * Usage: node scripts/normalise-reports.js [--dry-run] [--verbose]
 */

'use strict';

const fs   = require('fs');
const path = require('path');

const REPORTS_DIR = path.join(__dirname, '..', 'reports');
const TEMPLATE_PATH = path.join(__dirname, '..', 'reports', 'template.html');

// ── Canonical section map (normalize heading variations) ──────────────────────
const SECTION_MAP = {
  'executive summary':                        'Executive Summary',
  'business model':                           'Business Model',
  'financial snapshot':                       'Financial Snapshot',
  'financials':                               'Financial Snapshot',
  'recent catalysts':                         'Recent Catalysts',
  'recent catalysts (3-6 months)':            'Recent Catalysts',
  'catalysts':                               'Recent Catalysts',
  'thesis evaluation':                       'Thesis Evaluation',
  'bull / base / bear':                       'Thesis Evaluation',
  'scenario analysis':                       'Thesis Evaluation',
  'key risks':                                'Key Risks',
  'key risks (ranked)':                      'Key Risks',
  'risks':                                    'Key Risks',
  'who should own it / avoid it':            'Who Should Own It / Avoid It',
  'who should own it':                       'Who Should Own It / Avoid It',
  'ownership guidance':                      'Who Should Own It / Avoid It',
  'recommendation':                          'Recommendation',
  'entry / exit framework':                   'Entry / Exit Framework',
  'entry and exit':                           'Entry / Exit Framework',
  'price targets & risks':                   'Entry / Exit Framework',
  'sources':                                 'Sources',
};

// ── Canonical Sources template ─────────────────────────────────────────────────
const SOURCES_TEMPLATE = `
          <div class="report-section">
            <h2>Sources</h2>
            <ul>
              <li><strong>Authoritative market data:</strong> Live quote data sourced via DYOR HQ data pipeline (Google Sheets + Yahoo Finance). Fields sourced this way include price, market capitalisation, 52-week range, P/E, EPS, and volume.</li>
              <li><strong>Company filings and disclosures:</strong> Public company filings, regulatory announcements, investor presentations, and RNS where referenced in the analysis.</li>
              <li><strong>Additional public sources:</strong> Any third-party materials specifically cited inline in the report body.</li>
            </ul>
          </div>`;

// ── Biotech tickers requiring Paperclip pass ──────────────────────────────────
const BIOTECH_TICKERS = new Set([
  'AVCT', 'AVCT.L', 'ZYME', 'CRIS', 'AUTL', 'ABVE', 'C4X', 'C4X.L',
  'DESP', 'DESP.L', 'MRSN', 'POLXF', 'INO', 'MNTV', 'REGN', 'MRNA',
  'BMY', 'ABBV', 'ALKS', 'AMGN', 'GILD', 'EXAS', 'BMRN', 'CRNX',
  'DVAX', 'INDP', 'LMNL', 'OLMA', 'PCVX', 'PLRX', 'PRLR', 'RAPT',
  'RETA', 'SRPT', 'VTYX', 'XFOR',
]);

// ── Format B → Format A transformation ────────────────────────────────────────
const FORMAT_B_PATTERNS = [
  // Replace card-based structure with section-based
  { re: /<div class="card">\s*<div class="card-title">Executive Summary<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Executive Summary</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Business Model<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Business Model</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Financial Snapshot<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Financial Snapshot</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Recent Catalysts \(3-6 Months\)<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Recent Catalysts</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Thesis Evaluation<\/div>\s*<div class="scenario-grid">/g,
    into: '<div class="report-section"><h2>Thesis Evaluation</h2><div class="scenario-grid">' },
  { re: /<div class="card">\s*<div class="card-title">Key Risks \(Ranked\)<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Key Risks</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Who Should Own It \/ Avoid It<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Who Should Own It / Avoid It</h2><p>' },
  { re: /<div class="card">\s*<div class="card-title">Recommendation<\/div>\s*<p class="section-text">/g,
    into: '<div class="report-section"><h2>Recommendation</h2><p>' },
  // Close section divs (card → report-section)
  { re: /<\/p>\s*<\/div>\s*<!-- \/card -->\s*(?=<!-- (Business Model|Financial Snapshot|Recent Catalysts|Thesis Evaluation|Key Risks|Who Should Own|Recommendation|Sources|$))/g,
    into: '</p></div>' },
  // Replace card wrapper close with report-section close
  { re: /<\/div>\s*<!-- \/card -->/g, into: '</div>' },
  // Sources card → canonical Sources section
  { re: /<div class="card">\s*<div class="card-title">Sources<\/div>\s*<div class="source-item">([^<]+)<\/div>\s*<div class="source-item">([^<]+)<\/div>\s*<div class="source-item">([^<]*)<\/div>\s*<\/div>/g,
    into: SOURCES_TEMPLATE },
  // Closing body
  { re: /<\/div>\s*<\/div>\s*<\/div>\s*<\/main>\s*<footer>/g,
    into: '</div></div></div></div><aside class="report-sidebar"><div class="sidebar-card"><h3>Quick Stats</h3><div class="stat-row"><span>Ticker</span><span>{{TICKER}}</span></div></div></aside></div></main><footer>' },
];

// ── Helpers ───────────────────────────────────────────────────────────────────
function detectFormat(html) {
  if (html.includes('class="report-hero"'))        return 'A'; // canonical
  if (html.includes('class="ticker-badge"'))        return 'B'; // MDLZ-style card
  if (html.includes('<body>') && !html.includes('report-hero') && !html.includes('ticker-badge'))
    return 'C'; // generic placeholder
  return 'unknown';
}

function normaliseSectionHeadings(html) {
  let result = html;
  for (const [variant, canonical] of Object.entries(SECTION_MAP)) {
    const re = new RegExp(`<h2>\\s*${variant}\\s*</h2>`, 'gi');
    result = result.replace(re, `<h2>${canonical}</h2>`);
  }
  return result;
}

function ensureSourcesSection(html) {
  // If no Sources section exists, inject it before closing of report-content
  if (!html.includes('<h2>Sources</h2>')) {
    return html.replace(
      /(<div class="report-content">[\s\S]*?)(<\/div>\s*<aside class="report-sidebar">)/,
      `$1${SOURCES_TEMPLATE}$2`
    );
  }
  // If Sources exists but is in wrong format, normalise it
  return html;
}

function applyFormatBTransformation(html) {
  let result = html;
  for (const { re, into } of FORMAT_B_PATTERNS) {
    result = result.replace(re, into);
  }
  return result;
}

function extractTicker(filename) {
  return filename.split('-')[0].toUpperCase();
}

function isBiotech(ticker) {
  return BIOTECH_TICKERS.has(ticker) || BIOTECH_TICKERS.has(ticker.toUpperCase());
}

// ── Main ──────────────────────────────────────────────────────────────────────
const args = process.argv.slice(2);
const DRY_RUN   = args.includes('--dry-run');
const VERBOSE   = args.includes('--verbose');
const FORCE     = args.includes('--force');

const files = fs.readdirSync(REPORTS_DIR)
  .filter(f => f.endsWith('.html') && f !== 'template.html');

let updated = 0, skipped = 0, errors = 0;
const log = [];

for (const file of files.sort()) {
  const filepath = path.join(REPORTS_DIR, file);
  let html;

  try {
    html = fs.readFileSync(filepath, 'utf8');
  } catch(e) {
    errors++;
    console.error(`ERROR reading ${file}: ${e.message}`);
    continue;
  }

  const format = detectFormat(html);
  const ticker = extractTicker(file);

  if (VERBOSE) console.log(`${file}: format=${format}`);

  let normalised = html;
  let changed = false;

  // ── Step 1: Normalise section headings for all formats ──
  const beforeHeadings = normaliseSectionHeadings(normalised);
  if (beforeHeadings !== normalised) { normalised = beforeHeadings; changed = true; }

  // ── Step 2: Convert Format B → Format A ──
  if (format === 'B') {
    const converted = applyFormatBTransformation(normalised);
    if (converted !== normalised) { normalised = converted; changed = true; }
  }

  // ── Step 3: Ensure Sources section present and canonical ──
  const afterSources = ensureSourcesSection(normalised);
  if (afterSources !== normalised) { normalised = afterSources; changed = true; }

  // ── Step 4: Check for placeholder text in Business Model (Format C symptom) ──
  const bmMatch = normalised.match(/<h2>Business Model<\/h2>([\s\S]{10,200}?)<h2>/);
  if (bmMatch) {
    const bmText = bmMatch[1].replace(/<[^>]+>/g, '').toLowerCase();
    const isPlaceholder = bmText.includes('consumer discretionary sector company') ||
                         bmText.includes('sector company') ||
                         bmText.includes('investment case rests on') ||
                         bmText.includes('key drivers include');
    if (isPlaceholder) {
      log.push({ file, ticker, issue: 'PLACEHOLDER_TEXT', format });
      if (VERBOSE) console.log(`  WARNING: placeholder text detected`);
    }
  }

  if (!changed && !FORCE) {
    skipped++;
    if (VERBOSE) console.log(`  no changes`);
    continue;
  }

  if (DRY_RUN) {
    console.log(`DRY RUN: would update ${file} (format ${format}→A)`);
    continue;
  }

  try {
    fs.writeFileSync(filepath, normalised, 'utf8');
    updated++;
    log.push({ file, ticker, format, action: 'updated' });
    if (VERBOSE) console.log(`  updated`);
  } catch(e) {
    errors++;
    console.error(`ERROR writing ${file}: ${e.message}`);
  }
}

// ── Report ─────────────────────────────────────────────────────────────────────
console.log('\n── Normalisation Summary ──────────────────────────');
console.log(`Files processed: ${files.length}`);
console.log(`Updated:         ${updated}${DRY_RUN ? ' (dry run)' : ''}`);
console.log(`Unchanged:       ${skipped}`);
console.log(`Errors:          ${errors}`);

if (log.length) {
  console.log('\n── Issues ─────────────────────────────────────────');
  for (const e of log) {
    if (e.issue === 'PLACEHOLDER_TEXT') {
      console.log(`  ⚠   ${e.file} — placeholder text in Business Model (needs regen)`);
    } else if (e.action === 'updated') {
      console.log(`  ✓  ${e.file} — updated (format ${e.format}→A)`);
    }
  }
}

process.exit(errors > 0 ? 1 : 0);

File: scripts/normalise-reports.py
#!/usr/bin/env python3
"""
DYOR HQ v2 — Report Normalisation Script
Ensures every report follows the canonical template structure exactly.
Runs in place — modifies reports/*.html directly.
No commit until validated.
"""

import os, re, json

REPORTS_DIR = os.path.dirname(os.path.abspath(__file__)) + '/../reports'

# Canonical section name map (variant → canonical)
SECTION_MAP = {
    'executive summary':                          'Executive Summary',
    'business model':                              'Business Model',
    'financial snapshot':                          'Financial Snapshot',
    'recent catalysts (3-6 months)':             'Recent Catalysts',
    'catalysts':                                   'Recent Catalysts',
    'thesis evaluation':                           'Thesis Evaluation',
    'bull / base / bear':                          'Thesis Evaluation',
    'scenario analysis':                           'Thesis Evaluation',
    'key risks (ranked)':                          'Key Risks',
    'risks':                                       'Key Risks',
    'who should own it / avoid it':               'Who Should Own It / Avoid It',
    'ownership guidance':                          'Who Should Own It / Avoid It',
    'entry / exit framework':                      'Entry / Exit Framework',
    'entry and exit':                              'Entry / Exit Framework',
    'price targets & risks':                       'Entry / Exit Framework',
}

# Canonical Sources HTML
SOURCES_HTML = """
          <div class="report-section">
            <h2>Sources</h2>
            <ul>
              <li><strong>Authoritative market data:</strong> Live quote data sourced via DYOR HQ data pipeline (Google Sheets + Yahoo Finance). Fields sourced this way include price, market capitalisation, 52-week range, P/E, EPS, and volume.</li>
              <li><strong>Company filings and disclosures:</strong> Public company filings, regulatory announcements, investor presentations, and RNS where referenced in the analysis.</li>
              <li><strong>Additional public sources:</strong> Any third-party materials specifically cited inline in the report body.</li>
            </ul>
          </div>
"""

# Placeholder detection
PLACEHOLDER_PATTERNS = [
    re.compile(r'<h2>Business Model</h2>\s*<p>\s*(?:Consumer Discretionary|Consumer Staples|sector company|investment case rests on|Key drivers include)'),
    re.compile(r'Consumer Discretionary sector company'),
    re.compile(r'sector company\.?\s*</p>'),
]


def detect_format(html):
    if 'class="report-hero"' in html:
        return 'A'
    if 'class="ticker-badge"' in html:
        return 'B'
    return 'C'


def normalise_headings(html):
    result = html
    for variant, canonical in SECTION_MAP.items():
        pattern = re.compile(rf'<h2>\s*{re.escape(variant)}\s*</h2>', re.IGNORECASE)
        result = pattern.sub(f'<h2>{canonical}</h2>', result)
    return result


def convert_format_b(html):
    """Convert Format B (card/ticker-badge) → Format A (report-section)."""
    result = html

    # Remove body wrapper if present
    result = re.sub(r'^<body>\s*', '', result, flags=re.MULTILINE)
    result = re.sub(r'\s*</body>\s*$', '', result, flags=re.MULTILINE)

    # Replace .header-block with real header
    result = re.sub(
        r'<div class="header">\s*<div class="header-inner">\s*<div>\s*'
        r'<div class="ticker-badge">(\w+)</div>\s*'
        r'<div class="company-name">([^<]+)</div>',
        r'''<header>
    <div class="header-inner">
      <a href="../index.html" class="logo">
        <span class="logo-wordmark">DYOR <span>HQ</span></span>
        <span class="logo-badge">AI Research</span>
      </a>
      <nav>
        <ul class="nav-links">
          <li><a href="../index.html">← All Reports</a></li>
        </ul>
      </nav>
    </div>
  </header>
  <main>
    <div class="container">
      <div class="report-hero">
        <div class="report-breadcrumb">
          <a href="../index.html">Reports</a>
          <span>/</span>
          <span>\1</span>
        </div>
        <div class="report-title-row">
          <div class="report-title-block">
            <div class="ticker-label">\1</div>
            <h1>\2</h1>''',
        result
    )

    # Replace .card-title sections → report-section + h2
    section_replacements = [
        (r'<div class="card">\s*<div class="card-title">Executive Summary</div>',
         '<div class="report-section"><h2>Executive Summary</h2>'),
        (r'<div class="card">\s*<div class="card-title">Business Model</div>',
         '<div class="report-section"><h2>Business Model</h2>'),
        (r'<div class="card">\s*<div class="card-title">Financial Snapshot</div>',
         '<div class="report-section"><h2>Financial Snapshot</h2>'),
        (r'<div class="card">\s*<div class="card-title">Recent Catalysts \(3-6 Months\)</div>',
         '<div class="report-section"><h2>Recent Catalysts</h2>'),
        (r'<div class="card">\s*<div class="card-title">Thesis Evaluation</div>',
         '<div class="report-section"><h2>Thesis Evaluation</h2>'),
        (r'<div class="card">\s*<div class="card-title">Key Risks \(Ranked\)</div>',
         '<div class="report-section"><h2>Key Risks</h2>'),
        (r'<div class="card">\s*<div class="card-title">Who Should Own It / Avoid It</div>',
         '<div class="report-section"><h2>Who Should Own It / Avoid It</h2>'),
        (r'<div class="card">\s*<div class="card-title">Recommendation</div>',
         '<div class="report-section"><h2>Recommendation</h2>'),
        (r'<div class="card">\s*<div class="card-title">Entry / Exit Framework</div>',
         '<div class="report-section"><h2>Entry / Exit Framework</h2>'),
    ]
    for pattern, replacement in section_replacements:
        result = re.sub(pattern, replacement, result)

    # Replace section-text → clear para
    result = result.replace('<p class="section-text">', '<p>')

    # Close card divs and wrap in report-section
    result = re.sub(r'</p>\s*</div>\s*<!-- /card -->', '</p></div>', result)
    result = re.sub(r'</div>\s*<!-- /card -->', '</div>', result)

    # Inject sidebar after report-content close, before main close
    result = re.sub(
        r'(</div>\s*</div>\s*</div>\s*)(?=</div>\s*</div>\s*</div>\s*<footer>)',
        r'''\1        <aside class="report-sidebar">
          <div class="sidebar-card">
            <h3>Quick Stats</h3>
            <div class="stat-row"><span>Ticker</span><span>TICKER_PLACEHOLDER</span></div>
          </div>
        </aside>
        ''',
        result
    )

    return result


def ensure_sources(html):
    """Ensure canonical Sources section is present."""
    if '<h2>Sources</h2>' in html:
        return html
    # Inject before the closing of report-content div
    return re.sub(
        r'(<div class="report-content">[\s\S]+?)(</div>\s*<aside)',
        r'\1' + SOURCES_HTML + r'\2',
        html
    )


def check_placeholder(html):
    for pattern in PLACEHOLDER_PATTERNS:
        if pattern.search(html):
            return True
    return False


def normalise_one(filepath, dry_run=True, verbose=False):
    with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
        html = f.read()

    original = html
    fmt = detect_format(html)
    ticker = os.path.basename(filepath).split('-')[0].upper()

    changed = False

    # Step 1: normalise headings
    html = normalise_headings(html)
    if html != original: changed = True

    # Step 2: convert Format B
    if fmt == 'B':
        html = convert_format_b(html)
        if html != original: changed = True

    # Step 3: ensure Sources
    html = ensure_sources(html)
    if html != original: changed = True

    # Check placeholder
    has_placeholder = check_placeholder(html)

    if verbose or (not dry_run and changed):
        status = 'UPDATED' if changed else 'OK'
        placeholder_note = ' [PLACEHOLDER]' if has_placeholder else ''
        print(f'  [{status}] {os.path.basename(filepath)} (fmt {fmt}{placeholder_note})')

    if not dry_run and changed:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(html)

    return {'file': os.path.basename(filepath), 'fmt': fmt,
            'changed': changed, 'placeholder': has_placeholder}


def main():
    dry_run = '--dry-run' in __import__('sys').argv
    verbose = '--verbose' in __import__('sys').argv

    files = [f for f in os.listdir(REPORTS_DIR)
             if f.endswith('.html') and f != 'template.html']

    print(f'Processing {len(files)} reports '
          f'{"(DRY RUN)" if dry_run else "(WRITING)"}...\n')

    results = []
    for fn in sorted(files):
        results.append(normalise_one(os.path.join(REPORTS_DIR, fn), dry_run, verbose))

    updated   = sum(1 for r in results if r['changed'])
    unchanged = sum(1 for r in results if not r['changed'])
    placeholders = [r for r in results if r['placeholder']]

    print(f'\n── Summary ─────────────────────────────────────────')
    print(f'Total files:     {len(files)}')
    print(f'Updated:         {updated}')
    print(f'Unchanged:       {unchanged}')
    print(f'Placeholder:     {len(placeholders)} (need regen)')

    if placeholders:
        print(f'\n⚠  Reports needing regeneration:')
        for r in placeholders:
            print(f'   {r["file"]}')

    if not dry_run:
        print(f'\n✓  All changes written to disk.')
        print(f'   Review outputs before running: git diff reports/')


if __name__ == '__main__':
    main()